### PR TITLE
JP-3669: Updating the C Extension to do CHARGELOSS Read Noise Recalculations

### PR DESCRIPTION
**Pull request by kmacdonald-stsci**
Resolves [JP-3669](https://jira.stsci.edu/browse/JP-3669)

Closes #8601

This PR addresses the read noise variance recalculation based on the CHARGELOSS flag.  This recalculation was in the JWST ramp fit step code, but that came with a performance hit and was unnecessarily complicated.  The recalculation has been implemented in the C extension code, simplifying the ramp fitting, as well as improving performance.

There is a corresponding JWST PR that removes the CHARGELOSS recalculation from the step code, as well as removes the corresponding test. https://github.com/spacetelescope/jwst/pull/8678

The regression tests pass, except for one test.  The test failure has been confirmed to be incorrect.  The regression test truth file for this test needs to be updated:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1623/

**Checklist**

- [ ] added entry in `CHANGES.rst` (either in `Bug Fixes` or `Changes to API`)
- [ ] updated relevant tests
- [ ] updated relevant documentation
- [ ] updated relevant milestone(s)
- [ ] added relevant label(s)

### CODE REVIEW

File: CHANGES.rst
Comment (melanieclarke) [inline CHANGES.rst:13]:
```suggestion
  code to the C extension to simplify the code and improve performance. [#275]
```

File: src/stcal/ramp_fitting/ramp_fit.py
Comment (melanieclarke) [inline src/stcal/ramp_fitting/ramp_fit.py:62]: I have noticed in testing the previous version of ramp fit that this `where` call takes a ton of memory to complete, maybe because of the forced casting to np.uint32.  Since this is only used to decide whether to copy the groupdq here, can we find a less expensive substitute?  And maybe only call it if algorithm is OLS_C?

File: src/stcal/ramp_fitting/ramp_fit_class.py
Comment (melanieclarke) [inline src/stcal/ramp_fitting/ramp_fit_class.py:56]: Add orig_gdq to docstring

**Review by melanieclarke — COMMENTED** [review]
I'm not great at C, but this looks plausible to me.  Just a couple comments on the Python parts and some typos.

File: src/stcal/ramp_fitting/ramp_fit.py
Comment (kmacdonald-stsci) [reply src/stcal/ramp_fitting/ramp_fit.py:62]: I can research this.  Maybe `np.any` can be used in conjunction with `np.bitwise_and` to be faster than `np.where`.  Also, I will see if this is necessary.  The copy is done because the arrays get modified in during ramp fitting.  However, the way the C-extension is written, the code operates on local copies, possibly making this copy unnecessary.  In the mean time I conditioned it on running the `OLS_C` algorithm.

**Review by drlaw1558 — APPROVED** [review]
Linked to https://github.com/spacetelescope/jwst/pull/8697

Looks good to me!  I've successfully tested that:

1) Runtime improves as expected in test case jw02079004003_03101_00001_nis.  In the old version it takes about 2.5 minutes to run the relevant piece of the RN recalculation, while with this PR it takes about 2.5 seconds.

2) RN variances stay as expected in test case jw01510001001_02108_00001_nis using both the OLS and OLS_C algorithms.

3) Checked that the SCI and variance arrays did not change for other instruments that don't use charge migration.

File: CHANGES.rst
Comment (zacharyburnett) [inline CHANGES.rst:14]:
```suggestion
ramp_fitting
~~~~~~~~~~~~

- Move the CHARGELOSS read noise variance recalculation from the JWST step
  code to the C extension to simplify the code and improve performance.[#275]

```
#284 switched change log handling to `towncrier`, so this change log entry should be a file in `changes/`:
```shell
echo "[ramp_fitting] Move the CHARGELOSS read noise variance recalculation from the JWST step code to the C extension to simplify the code and improve performance." > changes/275.general.rst
```

(new PRs will have these instructions in the checklist)

**Review by melanieclarke — APPROVED** [review]
All my comments were addressed, thanks!

I think this should be good to go, once the changelog format is fixed and conflicts are resolved.  It might be a good idea to run the regression tests one last time before merging.

The corresponding JWST PR can be merged as soon as this one is in.

**Comment by tapastro** [comment]
This needs an update to move the changelog entry to a towncrier fragment, then we'll get this merged ASAP in preparation for an stcal build.

**Comment by tapastro** [comment]
Tests are showing a number of failures. It looks like the gls_fit tests have out of date dqflags, and the downstream tests show the same failures wrt. level 2 pipelines across PRs. We can get this PR in, but we should clean up these failures. There isn't much value gained by having tests if we ignore them due to repeated failures.

**Comment by kmacdonald-stsci** [comment]
> Tests are showing a number of failures. It looks like the gls_fit tests have out of date dqflags, and the downstream tests show the same failures wrt. level 2 pipelines across PRs. We can get this PR in, but we should clean up these failures. There isn't much value gained by having tests if we ignore them due to repeated failures.

I added the CHARGELOSS flag keyword to the GLS flags for testing.  They now pass.

**Comment by schlafly** [comment]
Is there an option to turn on/off this behavior?  For Roman we don't currently have this flag but we could add it; it does seem to me that we will have charge migration issues.  On the other hand, this code isn't about flagging or addressing the charge migration per se.  My understanding is that the "old" behavior gives a more accurate estimation of the read noise contribution to the variance of these pixels, but that we've decided that we don't want that more accurate estimation because it leads to high flux pixels being weighted downward in resample for IVM weighting.  I might have approached this via something like:
https://github.com/spacetelescope/jwst/pull/7563
where in resample one replaces the weights of saturated (or now also chargeloss'd) pixels with their unaffected neighbors, reasoning that one doesn't "really" care about the read noise variance for these pixels, where the poisson noise is dominant.

File: src/stcal/ramp_fitting/ramp_fit_class.py
Comment (braingram) [inline src/stcal/ramp_fitting/ramp_fit_class.py:147]: Is `CHARGELOSS` only supported by `OLS_C` and not `OLS`? If so, I think (and I tested this locally) passing in `algorithm` to this method and only setting `flags_chargeloss` if `algorithm == OLS_C` should fix the romancal tests.

Comment (kmacdonald-stsci) [reply src/stcal/ramp_fitting/ramp_fit_class.py:147]: The `CHARGELOSS` flag is used when selecting either `OLS_C` or `OLS`.  Currently, the use of the `CHARGELOSS` flag is handled in the `JWST` step code.  Unfortunately, there is a bug in the segmentation computation (that occurs very rarely on an edge case) and with a computation hit.  

When running the `OLS` code, there is a message that the code returned from the `STCAL` code, but there can be a noticeable lag between that and the completion of the ramp fitting step.  To fix the bug and to improve computational performance, I moved the read noise variance re-calculation from the `JWST` step code into the C-extension.  This required some accounting changes in `ramp_fit_class.py` and `ramp_fit.py` to get the proper information to the C-extension.

**Comment by nden** [comment]
I got confirmation from RTB that they do not intend to use OLS_C or OLS for Roman. 
If necessary, changes related to Roman will be done later. This can be merged  for JWST.

**Comment by kmacdonald-stsci** [comment]
> Happy to merge this ASAP, but it looks like one of the new unit tests is failing:
> 
> ```
> FAILED tests/test_ramp_fitting.py::test_cext_chargeloss - assert np.float32(0.000556237) != np.float32(0.000556237)
> ```

I"ll check it out.

**Comment by kmacdonald-stsci** [comment]
> Happy to merge this ASAP, but it looks like one of the new unit tests is failing:
> 
> ```
> FAILED tests/test_ramp_fitting.py::test_cext_chargeloss - assert np.float32(0.000556237) != np.float32(0.000556237)
> ```

It's fixed.

**Review by tapastro — APPROVED** [review]
LGTM - merging this momentarily.

**Pull request merged**
